Fuzzer fixes

yetamrra opened the pull request:
To compile with clang's LibFuzzer, the fuzzer target must be written in C++.  This requires including airscan.h into a C++ program.  The first three patches clean up compiler warnings and errors caused by doing so.

The last two patches add several checks for empty strings and NULL pointers that were turned up by initial runs of a fuzzer on the http_uri functions.

This is the first PR for #87.  I will send another with the fuzzer targets once we confirm that they're stable.

alexpevzner commented:
I've made mem_resize() type-safe (returning typed pointer instead of void*), so explicit cast is probably not needed anymore.

I also accepted your redefinition of error type, but using different name for the struct

Please, update your patch, then we will discuss other parts

alexpevzner commented:
Next question.

Your patch to `http_uri_parse_scheme` means that `http_uri_parse_scheme` could be called with `str == NULL`, which could only happen if `http_uri_new` or `http_uri_new_relative` is canned with `NULL` parameter, which actually should not happen.

Is it possible to somehow obtain an entire call path for this case?

And the same question for the `mem_trunc()` function, which is also not expected to be called with the `NULL` argument

yetamrra commented:
I dropped the patch that was adding casts, since your mem_resize change made it unnecessary.  In C++, you can't have a struct typedef with the same name as a pointer, so I renamed struct error to struct error_.  Is that name OK?

You're right that we're calling http_uri_new(NULL).  The fuzzer target takes whatever input the fuzzer produces and passes it to http_uri_new, and that was one of the cases it came up with.  I'll send another PR on Monday with that code so you can see the whole thing.

My basic idea is to try to eventually fuzz everything exposed in airscan.h.  Even though there might not be intended to be a direct path from the public sane calls to those functions, some of these paths might be reachable by invalid data coming back from a scanner.

alexpevzner commented:
If `http_uri_new(NULL)` is called directly by fuzzer (not as a side effect of calling some another function with valid input), I don't agree. The contract of this function doesn't allow to call it with NULL argument. I'm usually trying to be explicit in function's comments, is it allowed to be called with NULL argument, how the memory ownership is transferred and so on, but honestly, some details can be missed from my comments. Anyway, by default NULL is not allowed.

By default, all functions, that accept `char*`, assume that pointer is valid and string is `\0`-terminated, but doesn't assume that content of this string is valid. So it's OK (and makes sense) to test `http_uri_new` against any junk strings, but not against NULL pointers.

Anyway, if we "strengthen" all functions to accept NULL pointers, it will not actually make program to work more reliable. Instead, in the situation when, due to NULL pointer passing by mistake, not strengthened program will simple crash, so mistake will be quickly noticed and fixed, "strengthened" will pretend to work, somehow, but most likely the behaviour will be incorrect in some very inobvious way. This is like JavaScript programming, where almost every function accepts almost any input and doesn't crash, but when something goes wrong, you have no idea where to start searching for the reason of the problem.

yetamrra commented:
I think your position is reasonable when sane-airscan is used in a trusted environment and any bad data comes in by mistake.  But out in the wild, these functions are being exposed to data coming back from "random" untrusted and potentially malicious devices on the network.  I'm not arguing that the functions should accept a NULL pointer and attempt to do something reasonable with it, but they should return an error immediately rather than letting it propagate until it eventually crashes.

As a somewhat concrete example of getting to a NULL pointer without starting from an invalid string, consider 2522bf2fb08dfb5381c8a67ff7ba9038a5eb32d4 in this PR.  That commit is a fix for a crash that happens when `log_panic` tries to call `str_trunc` if you create a uri from `http://[A%2525]//MM` and pass it to `http_uri_fix_end_slash`.  It's clearly an invalid hostname, but the URI makes it past the parser.  The only code path that does that combination currently is in airscan-zerconf.c.  I briefly looked and didn't see a way to get that hostname through through the mDNS response, but that doesn't mean a more clever attacker than me couldn't find a way.  Even if there's no way to get into that specific path today, it's easy to imagine that a future bugfix or new feature might introduce a new way to call those functions, or there might be a different way to trigger a similar str_trunc elsewhere.

At the least, please consider sticking in a `log_assert` or an `abort` when NULL is passed to a function that shouldn't accept it; then the crash will happen as soon as the NULL is detected instead of at some later point that happens to dereference the pointer.

alexpevzner commented:
My position is following:
1. Invalid input data and internal program errors are not the same
2. No bits of input data could be trusted, and a reasonable recovery effort should be performed, if input data is invalid
3. Internal (programmer's) errors should cause program to crash loudly instead of recovery

Your example is very good and perfectly illustrates my idea. The real bug was in URI parser, thanks for finding it out! It simply should not accept invalid hosts like this. However, avoiding crash in this case would only hide the real error, it will not actually make the program stronger, only less observable,

Also please note, that logger assumes that `log_init` is called early, before any other call of any logger function. Adding initialization check into the logger with reporting panic via the logger itself makes things a way too tricky, lets avoid it.

I have no objection to add NULL checks via `log_assert`until a reasonable extent, but lets avoid it in trivial cases, it just makes code messy.

yetamrra commented:
Sorry, I missed the notification about your reply.  I see that you've already fixed the URI hostname parsing.  I'll clean this up and push a minimized version tomorrow.

alexpevzner commented:
I think that if you will post the following patch:

    MDNS_SERVICE    service = (MDNS_SERVICE)(intptr_t) userdata;

as a separate PR, I could accept it immediately, without any discussion. But please add a line of comment why it is needed, so at the next code refactoring I will not want to remove this type cast :-)

yetamrra commented:
Thanks for accepting the other PR.  I've simplified this one down to just a couple of non-NULL assertions at the main entry points.  Please let me know if that's the level of validation you think is appropriate.
